### PR TITLE
Add public constructor for RetroDevicePort.

### PR DESCRIPTION
**Pull request by InquisitiveCoder**
Addresses [Issue #6: Can't call RetroRuntime::is_joypad_button_pressed](https://github.com/libretro-rs/libretro-rs/issues/6).

### CODE REVIEW

File: libretro-rs/src/lib.rs [hunk 1]
```diff
@@ -110,6 +110,10 @@ impl TryFrom<u32> for RetroDevice {
 pub struct RetroDevicePort(u32);
 
 impl RetroDevicePort {
+  pub fn new(port_num: u32) -> Self {
+    RetroDevicePort(port_num)
+  }
```
Comment (adam-becker) [inline libretro-rs/src/lib.rs:115]: I looked briefly, but couldn't find if there's an upper limit on what a device port number can be? I'd like to enforce things like this if at all possible. Do you happen to know if there is one? The basic pattern would then be:

```rust
pub const UPPER_LIMIT: u32 = ???;

pub fn from_raw(val: u32) -> Option<Self> {
  if val < Self::UPPER_LIMIT {
    None
  } else {
    Some(Self(val))
  }
}
```

In lieu of an official number, perhaps adding a sufficiently large limit and getting the pattern in place ahead of time is the correct move?

Comment (InquisitiveCoder) [reply libretro-rs/src/lib.rs:115]: Since you asked, I searched up and down libretro.h and also did some searches on the Libretro Discord and came out empty-handed.

To give a sense of scale, RetroArch defaults to 5 ports but can be changed in the UI settings to go as low as 1 or as high as 16. The latter is currently a hard-coded limit; [there's a widely-used constant for it in the code](https://github.com/libretro/RetroArch/blob/b403f5fc1fb78885c8898cf9b94f521dd4cfb70d/input/input_defines.h#L27). But that's just one frontend; others could have different limits. Case in point, [I just found this frontend called Ludo that seems to have a hard limit of 5](https://github.com/libretro/ludo/blob/master/input/input.go). A core ought to function properly regardless of which frontend is driving it.

So as best as I can tell the expectation is that 1) cores will self-impose some limit that makes sense for their use case, 2) cores won't make assumptions about the number of ports a frontend provides, 3) frontends won't make assumptions about the number of ports the core will try to query and 4) cores will gracefully handle ports being set to `RETRO_DEVICE_NONE`.

My 2 cents is that I can't imagine a use case for more than 16 ports but I'd be uncomfortable with this crate setting limits that aren't part of the libretro spec just as a matter of principle. If someone really wants to do that they shouldn't be forced to abandon the crate or code in C/C++.

Comment (adam-becker) [reply libretro-rs/src/lib.rs:115]: I'm fine with imposing a 16 limit for now. I'm hard pressed to think of games that support more than 4 or 8 in local coop, so 16 seems reasonable.

Comment (InquisitiveCoder) [reply libretro-rs/src/lib.rs:115]: I still think it's a mistake to set the limit at 16; it's an implementation detail that RetroArch could change at any time. The libretro API doesn't provide any way to determine what the frontend's limit is; that implies frontends need to be able to gracefully handle port numbers beyond their self-imposed limit. Case in point, I modified my core to poll gamepad button A on port 17 and RetroArch handled it just fine.

Conversely, RetroArch doesn't always provide 16 ports; it defaults to 5 and could be configured to use as few as 1. Any multiplayer core needs to handle frontends with fewer ports than the core's maximum. A 20 port core would still need to be written to handle as few as 1 port.

That said, I did find that at least on x64 Windows, RetroArch has a hard limit of 65601. I'm able to poll up to port 65600, but polling port 65601 causes a segfault. 65601 is also `u16::MAX + 66`, so we should probably limit the port number to `u16`.

Comment (InquisitiveCoder) [reply libretro-rs/src/lib.rs:119]: I looked further into how [Ludo](https://github.com/libretro/ludo) (which seems to be the main alternative to RetroArch) handles input and [it also seems to handle polling ports beyond their maximum gracefully](https://github.com/libretro/ludo/blob/819878fdbb872d99f5432c87ea0b1aff3447edf5/input/input.go#L177):
```go
if port >= MaxPlayers {
    return 0
}
```

File: libretro-rs/src/lib.rs [hunk 2]
```diff
@@ -106,16 +107,27 @@ impl TryFrom<u32> for RetroDevice {
   }
 }
 
-#[derive(Clone, Copy, Debug)]
-pub struct RetroDevicePort(u32);
+/// A libretro device port.
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RetroDevicePort(c_uint);
```
Comment (adam-becker) [inline libretro-rs/src/lib.rs:113]: Why the change to `c_uint`?

Comment (InquisitiveCoder) [reply libretro-rs/src/lib.rs:113]: `RetroInstance::on_set_controller_port_device` is going to create a RetroDevicePort out of a `c_uint` using the private constructor. Shouldn't we ensure ensure the types always match on every architecture?

Comment (adam-becker) [reply libretro-rs/src/lib.rs:113]: Translations of this kind should be done in the glue layer, everything user-facing should be a safe, known-size, rust type in my opinion. Using `c_uint` here would cause `RetroDevicePort` to be a different size on different architectures.

In a worse case, it might force us to expose our `libc::c_uint` if we aren't very careful about API design. I think a good "rule of thumb" here is that all translations from the C API to Rust happen in the `on_*` methods of `RetroInstance` (either directly, or by calling an appropriate conversion on a struct, say `RetroDevicePort::try_from_raw` before we pass that data along to the trait. The structs themselves, however, should just have `u32` or whatever equivalent rust type.

Comment (InquisitiveCoder) [reply libretro-rs/src/lib.rs:113]: Considering Rust doesn't have implicit numeric coercions like C or a stable ABI, is the architecture-specific size really a problem? Even safe Rust has to deal with architecture-specific `usize` right?

I think we're on the same page regarding hiding as much of the C API's ugliness as possible (e.g. no raw pointers, newtypes instead of plain integer types, enforce safety by design) but I don't see how we can work around the fact that various parts of the API are architecture-specific without surfacing that to the end user, especially in this case where the core's both sending and receiving port numbers.

The way I look at it, the `c_uint` is win/win. If `c_uint = u32` in every architecture, the change makes no difference. But if there's cases where `c_uint != u32`, sticking with `u32` will either force us to 1) cause a compilation error, 2) cast at runtime and pray we never cast a value that'd overflow, or 3) check for overflow and panic. All of these seem strictly worse than surfacing the architecture-specific size to the user. That said I've never had to write FFI code or maintain a crate before so it's possible there's something I've overlooked? I'm not trying to twist your arm, I genuinely want to understand what makes the fixed size approach better so I can be better informed.

As for `libc`, I think we can remove that dependency altogether. Bindgen isn't generating `libc` types in `libretro.rs`, it's using `std::os::raw`. Since the crate still compiles when mixing `libc` and `std::os::raw` types, they must be using the same type aliases? Actually, the crate and its users should probably both be using the `core::ffi` types, which were copied over from `std::os::raw` so they could be used in `no_std` environments. I believe the types in `std::os::raw` now alias the ones in `core::ffi`, and the documentation for `std::os::raw` tells you to prefer the `core::ffi` module instead.

Comment (InquisitiveCoder) [reply libretro-rs/src/lib.rs]: After sleeping on it (literally) I realized there's no problem using a `u8` after all. Since we're limiting cores to using ports 0-255, if the frontend calls `retro_set_controller_device_port` with port 256+, there's nothing useful the core can do with that info and we can just ignore the function call.

**Comment by adam-becker** [comment]
@InquisitiveCoder will merge this tonight after work. Thanks again for the contributions!

**Comment by InquisitiveCoder** [comment]
My pleasure! I'm learning a lot and streamlining my core at the same time 👍
